src/transform/src/dataflow.rs

```diff
@@ -34,6 +37,131 @@ pub fn optimize_dataflow(dataflow: &mut DataflowDesc) {
+            if dataflow.objects_to_build[index].typ.is_some() {
+                let other = occurrences_in_later_views[0];
+                // We can remove this view and insert it in the later view,
+                // but are not able to relocate the later view `other`.
+
+                // When splicing in the `index` view, we need to create disjoint
+                // identifiers for the Let's `body` and `value`, as well as a new
+                // identifier for the binding itself. Following `UpdateLet`, we
+                // go with the binding first, then the value, then the body.
+                let update_let = crate::update_let::UpdateLet;
+                let mut id_gen = crate::IdGen::default();
+                let new_local = LocalId::new(id_gen.allocate_id());
+                // Use the same `id_gen` to assign new identifiers to `index`.
+                update_let.action(
+                    dataflow.objects_to_build[index].relation_expr.as_mut(),
+                    &mut HashMap::new(),
+                    &mut id_gen,
+                );
+                // Assign new identifiers to the other relation.
+                update_let.action(
+                    dataflow.objects_to_build[other].relation_expr.as_mut(),
+                    &mut HashMap::new(),
+                    &mut id_gen,
+                );
+                // Install the `new_local` name wherever `global_id` was used.
+                dataflow.objects_to_build[other]
+                    .relation_expr
+                    .as_mut()
+                    .visit_mut(&mut |expr| {
+                        if let RelationExpr::Get { id, .. } = expr {
+                            if id == &Id::Global(global_id) {
+                                *id = Id::Local(new_local);
+                            }
+                        }
+                    });
+
+                // With identifiers rewritten, we can replace `other` with
+                // a `RelationExpr::Let` binding, whose value is `index` and
+                // whose body is `other`.
+                let body = dataflow.objects_to_build[other]
+                    .relation_expr
+                    .as_mut()
+                    .take_dangerous();
+                let value = dataflow.objects_to_build[index]
+                    .relation_expr
+                    .as_mut()
+                    .take_dangerous();
+                *dataflow.objects_to_build[other].relation_expr.as_mut() = RelationExpr::Let {
+                    id: new_local,
+                    value: Box::new(value),
+                    body: Box::new(body),
+                };
+                dataflow.objects_to_build.remove(index);
+            }
+        }
+    }
+
+    // Re-optimize each dataflow.
+    // TODO: Unfortunately, we only have access to the indexes that are included in
+    // the dataflow. There could be better indexes to use, could we return to them.
```

Comment:
I just took a look at this, and while it seems not hard to provide the arguments and re-optimize, it was less clear what to do about arrangements that end up used. Like, we will need to add them to imports, maybe remove imports that aren't used, etc. It seems like maybe the imports should be determined from the final optimized representation, rather than manually added, but that this would be more involved and error prone.

Reply:
Yeah, perhaps imports should be determined from final optimized representation.

Just change the TODO to explain that we could plug in the indexes from the catalog, but it gets tangly adding new imports into the DataflowDesc and removing ones no longer used.